File: src/assertions/inventory.ts
import type { Agent } from '../core/client.js';
import type { InventoryItem } from '../types/index.js';
import { AssertionError } from './index.js';

export class InventoryAssertion {
  constructor(private player: Agent) {}

  /**
   * 指定アイテムを持っているか確認
   */
  toHaveItem(itemId: string, options?: { slot?: number }): this {
    const items = this.player.getInventory();
    const normalizedId = itemId.startsWith('minecraft:') ? itemId : `minecraft:${itemId}`;

    const found = options?.slot !== undefined
      ? items.find((item) => item.slot === options.slot && item.id === normalizedId)
      : items.find((item) => item.id === normalizedId);

    if (!found) {
      throw new AssertionError(
        `Expected inventory to contain "${normalizedId}"${options?.slot !== undefined ? ` in slot ${options.slot}` : ''}`,
        normalizedId,
        items.map((i) => i.id)
      );
    }
    return this;
  }

  /**
   * 指定アイテムを指定個数以上持っているか確認
   */
  toHaveItemCount(itemId: string, minCount: number): this {
    const items = this.player.getInventory();
    const normalizedId = itemId.startsWith('minecraft:') ? itemId : `minecraft:${itemId}`;

    const totalCount = items
      .filter((item) => item.id === normalizedId)
      .reduce((sum, item) => sum + item.count, 0);

    if (totalCount < minCount) {
      throw new AssertionError(
        `Expected at least ${minCount} of "${normalizedId}", but found ${totalCount}`,
        minCount,
        totalCount
      );
    }
    return this;
  }

  /**
   * 指定アイテムを持っていないことを確認
   */
  notToHaveItem(itemId: string): this {
    const items = this.player.getInventory();
    const normalizedId = itemId.startsWith('minecraft:') ? itemId : `minecraft:${itemId}`;

    const found = items.find((item) => item.id === normalizedId);

    if (found) {
      throw new AssertionError(
        `Expected inventory not to contain "${normalizedId}", but found ${found.count} in slot ${found.slot}`,
        undefined,
        found
      );
    }
    return this;
  }

  /**
   * インベントリが空かどうか確認
   */
  toBeEmpty(): this {
    const items = this.player.getInventory();

    if (items.length > 0) {
      throw new AssertionError(
        `Expected inventory to be empty, but found ${items.length} items`,
        0,
        items.length
      );
    }
    return this;
  }

  /**
   * インベントリが空でないことを確認
   */
  notToBeEmpty(): this {
    const items = this.player.getInventory();

    if (items.length === 0) {
      throw new AssertionError(
        'Expected inventory not to be empty',
        'items',
        'empty'
      );
    }
    return this;
  }

  /**
   * 指定スロットにアイテムがあるか確認
   */
  toHaveItemInSlot(slot: number): this {
    const items = this.player.getInventory();
    const found = items.find((item) => item.slot === slot);

    if (!found) {
      throw new AssertionError(
        `Expected item in slot ${slot}, but slot is empty`,
        'item',
        undefined
      );
    }
    return this;
  }

  /**
   * 指定エンチャントを持つアイテムがあるか確認
   */
  toHaveEnchantedItem(itemId: string, enchantmentId: string, minLevel?: number): this {
    const items = this.player.getInventory();
    const normalizedItemId = itemId.startsWith('minecraft:') ? itemId : `minecraft:${itemId}`;
    const normalizedEnchId = enchantmentId.startsWith('minecraft:') ? enchantmentId : `minecraft:${enchantmentId}`;

    const item = items.find((i) => i.id === normalizedItemId);
    if (!item) {
      throw new AssertionError(
        `Expected inventory to contain "${normalizedItemId}"`,
        normalizedItemId,
        undefined
      );
    }

    const enchantment = item.enchantments?.find((e) => e.id === normalizedEnchId);
    if (!enchantment) {
      throw new AssertionError(
        `Expected "${normalizedItemId}" to have enchantment "${normalizedEnchId}"`,
        normalizedEnchId,
        item.enchantments?.map((e) => e.id)
      );
    }

    if (minLevel !== undefined && enchantment.level < minLevel) {
      throw new AssertionError(
        `Expected enchantment "${normalizedEnchId}" to be at least level ${minLevel}, but was level ${enchantment.level}`,
        minLevel,
        enchantment.level
      );
    }

    return this;
  }

  /**
   * インベントリ変更を待機
   */
  async toReceiveItem(
    itemId: string,
    options?: { timeout?: number }
  ): Promise<InventoryItem> {
    const { timeout = 5000 } = options ?? {};
    const normalizedId = itemId.startsWith('minecraft:') ? itemId : `minecraft:${itemId}`;

    try {
      const [item] = await this.player.waitFor('inventory_update', {
        timeout,
        filter: (update: { item: InventoryItem }) => update.item.id === normalizedId,
      });
      return item.item;
    } catch {
      throw new AssertionError(
        `Timeout waiting for item "${normalizedId}" to be received`,
        normalizedId,
        undefined
      );
    }
  }
}

File: pkg/assertions/inventory.go
package assertions

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/gollilla/best/pkg/events"
	"github.com/gollilla/best/pkg/types"
)

// InventoryAssertion provides inventory-related assertions
type InventoryAssertion struct {
	agent AgentInterface
}

// ToHaveItem checks if the inventory contains a specific item
// itemID can be a full ID (e.g., "minecraft:diamond") or a partial match (e.g., "diamond")
func (i *InventoryAssertion) ToHaveItem(itemID string) {
	items := i.agent.GetInventory()

	for _, item := range items {
		if matchesItemID(item.ID, itemID) {
			return
		}
	}

	panic(NewAssertionError(
		fmt.Sprintf("expected inventory to have item %q", itemID),
		itemID,
		getInventoryItemIDs(items),
	))
}

// ToHaveItemCount checks if the inventory contains a specific count of an item
func (i *InventoryAssertion) ToHaveItemCount(itemID string, expectedCount int32) {
	items := i.agent.GetInventory()

	var totalCount int32
	for _, item := range items {
		if matchesItemID(item.ID, itemID) {
			totalCount += item.Count
		}
	}

	if totalCount == expectedCount {
		return
	}

	panic(NewAssertionError(
		fmt.Sprintf("expected inventory to have %d of item %q, but found %d", expectedCount, itemID, totalCount),
		expectedCount,
		totalCount,
	))
}

// ToHaveAtLeast checks if the inventory contains at least a certain count of an item
func (i *InventoryAssertion) ToHaveAtLeast(itemID string, minCount int32) {
	items := i.agent.GetInventory()

	var totalCount int32
	for _, item := range items {
		if matchesItemID(item.ID, itemID) {
			totalCount += item.Count
		}
	}

	if totalCount >= minCount {
		return
	}

	panic(NewAssertionError(
		fmt.Sprintf("expected inventory to have at least %d of item %q, but found %d", minCount, itemID, totalCount),
		minCount,
		totalCount,
	))
}

// ToBeEmpty checks if the inventory is empty
func (i *InventoryAssertion) ToBeEmpty() {
	items := i.agent.GetInventory()

	if len(items) == 0 {
		return
	}

	panic(NewAssertionError(
		"expected inventory to be empty",
		0,
		len(items),
	))
}

// ToReceiveItem waits for an inventory update containing a specific item
func (i *InventoryAssertion) ToReceiveItem(ctx context.Context, itemID string) *types.InventoryItem {
	data, err := i.agent.Emitter().WaitFor(ctx, events.EventInventoryUpdate, func(d events.EventData) bool {
		items, ok := d.([]types.InventoryItem)
		if !ok {
			return false
		}

		for _, item := range items {
			if matchesItemID(item.ID, itemID) {
				return true
			}
		}
		return false
	})

	if err != nil {
		panic(err)
	}

	items := data.([]types.InventoryItem)
	for _, item := range items {
		if matchesItemID(item.ID, itemID) {
			return &item
		}
	}

	panic(NewAssertionError(
		fmt.Sprintf("received inventory update but item %q not found", itemID),
		itemID,
		nil,
	))
}

// ToReceiveItemInSlot waits for an inventory slot update for a specific slot
func (i *InventoryAssertion) ToReceiveItemInSlot(ctx context.Context, slot int32) *types.InventoryItem {
	data, err := i.agent.Emitter().WaitFor(ctx, events.EventInventorySlotUpdate, func(d events.EventData) bool {
		item, ok := d.(types.InventoryItem)
		if !ok {
			return false
		}
		return item.Slot == slot
	})

	if err != nil {
		panic(err)
	}

	item := data.(types.InventoryItem)
	return &item
}

// WaitForInventoryChange waits for any inventory change within the timeout
func (i *InventoryAssertion) WaitForInventoryChange(timeout time.Duration) {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	_, err := i.agent.Emitter().WaitFor(ctx, events.EventInventoryUpdate, nil)
	if err != nil {
		panic(err)
	}
}

// Helper functions

// matchesItemID checks if an item ID matches the expected pattern
// Supports:
// - Full IDs (minecraft:diamond, item:335)
// - Partial matches (diamond matches minecraft:diamond)
// - Network IDs (335 matches item:335)
func matchesItemID(actualID, expectedID string) bool {
	// Exact match
	if actualID == expectedID {
		return true
	}

	// Check if expectedID is a numeric ID (e.g., "335" matches "item:335")
	if strings.HasPrefix(actualID, "item:") && strings.TrimPrefix(actualID, "item:") == expectedID {
		return true
	}

	// Partial match (e.g., "diamond" matches "minecraft:diamond")
	if strings.Contains(actualID, expectedID) {
		return true
	}

	// Check if expected has namespace but actual doesn't
	if strings.Contains(expectedID, ":") {
		parts := strings.Split(expectedID, ":")
		if len(parts) == 2 && strings.Contains(actualID, parts[1]) {
			return true
		}
	}

	return false
}

// getInventoryItemIDs returns a list of item IDs in the inventory
func getInventoryItemIDs(items []types.InventoryItem) []string {
	ids := make([]string, 0, len(items))
	for _, item := range items {
		ids = append(ids, item.ID)
	}
	return ids
}
